pyDrop/clustering/course_graining.py

```diff
@@ -0,0 +1,249 @@
+"""
+=========================================================================================
+                                 CALICO CLUSTERING
+=========================================================================================
+(Iterative clustering assisted by coarse-graining) developed in the following paper
+doi: 10.1021/acs.analchem.7b02688 
+Purpose: Increases the ability of unsupervised models to capture smaller clusters through 
+    course-graining. Iteratively course-grains the data the updated the starting location 
+    of the K-means clustering algorithm
+"""
+from sklearn.cluster import KMeans, OPTICS, DBSCAN, Birch, AgglomerativeClustering
+from sklearn.metrics.cluster import rand_score, homogeneity_score, completeness_score, v_measure_score
+import numpy as np
+import matplotlib.pyplot as plt
+from copy import deepcopy
+
+class ModuloBins:
+    """
+    ModuloBins
+    """
+    def __init__(self, mod=100, rem=0):
+        self.mod = mod
+        self.rem = rem
+
+        # Bin finding/evaluating functions to be called with np.arrays
+        self.id_to_bin_start = np.vectorize(self._id_to_bin_start)
+        self.id_to_bin_center = np.vectorize(self._id_to_bin_center)
+        self.value_to_id = np.vectorize(self._value_to_id)
+
+    def _id_to_bin_start(self, idx):
+        return float(idx*self.mod + self.rem)
+    
+    def _id_to_bin_center(self, idx):
+        return float(idx*self.mod + self.rem) + self.mod/2
+    
+    def _value_to_id(self, value):
+        return int((value-self.rem)/self.mod)
+    
+class LinSpaceBins:
+    """
+    LinSpaceBins
+    """
+    def __init__(self, min, max, n_bins=100):
+        self.min = min
+        self.max = max
+        self.n_bins = n_bins
+
+        self.bins = np.linspace(self.min, self.max, self.n_bins)
+        self.step = self.bins[1] - self.bins[0]
+
+        # Bin finding/evaluating functions to be called with np.arrays
+        self.id_to_bin_start = np.vectorize(self._id_to_bin_start)
+        self.id_to_bin_center = np.vectorize(self._id_to_bin_center)
+        self.value_to_id = np.vectorize(self._value_to_id)
+    
+    def id_to_bin_start(self, idx):
+        return self.bins[idx]
+    
+    def id_to_bin_center(self, idx):
+        return self.bins[idx] + self.step/2
+    
+    def value_to_id(self, value):
```

Comment:
Question: Shouldn't we be able to find the index of a bin via `value // self.step`? What you're doing here isn't wrong, but must equate to a binary search for every call?

Reply:
I do appreciate the simplicity this brings for floating point bin sizes / values though. If you were to use the `value // self.step` above, things can get tricky when the value is negative and around 0.0.

So after thinking about this a bit more, I'd go with what you have unless performance is an issue.

Reply:
Yeah that's a good point. It wouldn't take too much work to scale the data range by min and max to deal with negative / close-to-zero values. I think I might try this!